### PR TITLE
Ensure no logical errors on counter overflow

### PR DESCRIPTION
**Pull request by Dmitry-Me**
This resolves https://github.com/dotnet/corefx/issues/5414

### CODE REVIEW

File: src/System.Data.SqlClient/src/System/Data/ProviderBase/DbConnectionFactory.cs [hunk 1]
```diff
@@ -214,7 +214,11 @@ internal bool TryGetConnection(DbConnection owningConnection, TaskCompletionSour
                             // if didn't find one, pick the next one in round-robbin fashion
                             if (idx == s_pendingOpenNonPooled.Length)
                             {
-                                idx = s_pendingOpenNonPooledNext++ % s_pendingOpenNonPooled.Length;
+                                idx = s_pendingOpenNonPooledNext % s_pendingOpenNonPooled.Length;
```
Comment (JonHanna) [inline src/System.Data.SqlClient/src/System/Data/ProviderBase/DbConnectionFactory.cs]: Did this compile for you locally? Should there not be an explicit cast here?

**Comment by stephentoub** [comment]
LGTM

cc: @saurabh500

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by saurabh500** [comment]
LGTM

**Pull request merged**